### ISSUE TITLE
[no-array-callback-reference] enhancement: report only if missing mandatory arguments

### ISSUE THREAD

**Issue by yrtimiD** [sindresorhus/eslint-plugin-unicorn#1195]
I propose to add some flexibility to this rule by optionally allowing callbacks which already have some or all of the required arguments.

Current rule description example:
> The unicorn module now does a minor version that adds another argument: 
module.exports = (x, y) => x + (y ? y : 1);
Your code will now return something different and probably break for users because it is now passing the index of the item as second argument.

is absolutely valid, but if callback function already has all the required arguments - it's not danger anymore:
```js
function callback(value, index, array) { }
[1,2,3].map(callback); // should pass
```
since 3rd 'array' argument is rarely used, we can also allow functions with value and index arguments only.

to summarize:
```js
function c_noArgs() { }
function c_value(value) { }
function c_valueIndex(value, index) { }
function c_valueIndexArray(value, index, array) { }

[1,2,3].map(c_noArgs); // should fail
[1,2,3].map(c_value); // should fail
[1,2,3].map(c_valueIndex); // should pass if "AllowValueIndex: true"
[1,2,3].map(c_valueIndexArray); // should pass if "AllowValueIndexArray": true
```

**Comment by sindresorhus** [comment]
> is absolutely valid, but if callback function already has all the required arguments - it's not danger anymore:

It is if the callback function is not defined in the same scope. It's easy to accidentally change it without realizing it affects the call site.

We could maybe consider adding an option to allow this if the callback function is defined in the same scope or a parent scope.

**Comment by fregante** [comment]
More context here: https://jakearchibald.com/2021/function-callback-risks/

I think it's still risky. Let's say you have a project with a pre-existing bug:

```js
function c_valueIndex(value, notAnIndex) {
	if (notAnIndex) {
		return value * 2;
	}
}
[1,2,3].map(c_valueIndex); // should pass if "AllowValueIndex: true"
```

It wouldn't be caught by this exception.

The only situation that could **definitelytotally** be exempt is one in which we can guarantee that `arguments[1]` is expected to be an index. Can that even be done? Maybe it should literally be called `index` or `(.*)index` and nothing else?

At the same time though… should _all_ raw callbacks be avoided for this reason? 🤔 Why is `addEventListener(callback)` ok?

**Comment by sindresorhus** [comment]
Closing. This rule intentionally reports direct callback references, including locally declared callbacks. The safe pattern is to use an inline wrapper so the callback arguments are explicit.

Allowing callbacks based on arity would weaken the rule and still would not prove that the function was designed for the array callback contract.
